### ISSUE TITLE
Must use destructuring state assignment error for accessing variable state keys

### ISSUE THREAD

**Issue by jsjain** [yannickcr/eslint-plugin-react#2407]
I am trying to access `variable key` from my react state.
like: `if (typeof this.state[field] === 'object'))`

while doing so I am getting the error `Must use destructuring state assignmenteslint(react/destructuring-assignment)` 
In this case i wont be able to de structure the state. Is this something frowned upon ?

**Comment by ljharb** [comment]
```jsx
const { [field]: something } = this.props;
```

**Comment by jsjain** [comment]
Didn't knew we de-structure variable field like this as well.
Thanks @ljharb for the help.

**Comment by Emiya0306** [comment]
@ljharb 
Hi~ If I would like to write a component that just what [this blog](https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html#recommendation-fully-uncontrolled-component-with-a-key) saying.
The draft input component has initial state with props, and the state using class fields syntax. How to   destruct it?

```javascript
class EmailInput extends Component {
  state = { email: this.props.defaultEmail };

  handleChange = event => {
    this.setState({ email: event.target.value });
  };

  render() {
    return <input onChange={this.handleChange} value={this.state.email} />;
  }
}
```

**Comment by ljharb** [comment]
Initial state with props is incorrect unless it’s also using getDerivedStateFromProps.

Functions should not go in class fields, so you want this:
```jsx

class EmailInput extends Component {
  state = {};

  static getDerivedStateFromProps({ defaultEmail }) {
    return { email: defaultEmail };
  }

  handleChange = this.handleChange.bind(this);
  handleChange(event) {
    this.setState({ email: event.target.value });
  }

  render() {
    return <input onChange={this.handleChange} value={this.state.email} />;
  }
}

```

**Comment by Emiya0306** [comment]
Wow, that's cool. Thanks for help! 👍

**Comment by info-bit** [comment]
@ljharb "Functions should not go in class fields" why precisely?

**Comment by ljharb** [comment]
@info-bit conceptually, own properties are for data; functions are not data; functions should be shared across all instances. also, you can't spy/mock out functions in fields.
